File: Delta/UIDevice+Vibration.m
//
//  UIDevice+Vibration.m
//  Delta
//
//  Created by Riley Testut on 7/5/15.
//  Copyright © 2015 Riley Testut. All rights reserved.
//

#import "UIDevice+Vibration.h"

#include <sys/sysctl.h>

@import AudioToolbox;

// Private vibration API declarations
#if !APP_STORE_BUILD
void AudioServicesStopSystemSound(int);
void AudioServicesPlaySystemSoundWithVibration(int, id, NSDictionary *);
#endif

@implementation UIDevice (Vibration)

#pragma mark - Vibration -

- (void)vibrate
{
#if !APP_STORE_BUILD
    AudioServicesStopSystemSound(kSystemSoundID_Vibrate);
    
    int64_t vibrationLength = 30;
    
    if ([[self modelGeneration] hasPrefix:@"iPhone6"])
    {
        // iPhone 5S has a weaker vibration motor, so we vibrate for 10ms longer to compensate
        vibrationLength = 40;
    }
    
    NSArray *pattern = @[@NO, @0, @YES, @(vibrationLength)];
    
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    dictionary[@"VibePattern"] = pattern;
    dictionary[@"Intensity"] = @1;
    
    AudioServicesPlaySystemSoundWithVibration(kSystemSoundID_Vibrate, nil, dictionary);
#endif
}

#pragma mark - Device Info -

- (NSString *)modelGeneration
{
    char *type = "hw.machine";
    
    size_t size;
    sysctlbyname(type, NULL, &size, NULL, 0);
    
    char *name = malloc(size);
    sysctlbyname(type, name, &size, NULL, 0);
    
    NSString *modelName = [NSString stringWithCString:name encoding:NSUTF8StringEncoding];
    free(name);
    
    return modelName;
}

#pragma mark - Getters/Setters -

- (BOOL)isVibrationSupported
{
#if APP_STORE_BUILD || TARGET_IPHONE_SIMULATOR
    return NO;
#else
    // No way to detect if hardware supports vibration, so we assume if it's not an iPhone, it doesn't have a vibration motor
    return [self.model hasPrefix:@"iPhone"];
#endif
}

@end

File: Cores/DeltaCore/Sources/Extensions/UIDevice+Vibration.swift
//
//  UIDevice+Vibration.swift
//  DeltaCore
//
//  Created by Riley Testut on 11/28/16.
//  Copyright © 2016 Riley Testut. All rights reserved.
//

import UIKit
import AudioToolbox

@_silgen_name("AudioServicesStopSystemSound")
func AudioServicesStopSystemSound(_ soundID: SystemSoundID)

// vibrationPattern parameter must be NSDictionary to prevent crash when bridging from Swift.Dictionary.
@_silgen_name("AudioServicesPlaySystemSoundWithVibration")
func AudioServicesPlaySystemSoundWithVibration(_ soundID: SystemSoundID, _ idk: Any?, _ vibrationPattern: NSDictionary)

public extension UIDevice
{
    enum FeedbackSupportLevel: Int
    {
        case unsupported
        case basic
        case feedbackGenerator
    }
}

public extension UIDevice
{
    var feedbackSupportLevel: FeedbackSupportLevel
    {
        guard let rawValue = self.value(forKey: "_feedbackSupportLevel") as? Int else { return .unsupported }
        
        let feedbackSupportLevel = FeedbackSupportLevel(rawValue: rawValue)
        return feedbackSupportLevel ?? .feedbackGenerator // We'll assume raw values greater than 2 still support UIFeedbackGenerator ¯\_(ツ)_/¯
    }
    
    var isVibrationSupported: Bool {
        #if (arch(i386) || arch(x86_64))
            // Return false for iOS simulator
            return false
        #else
            // All iPhones support some form of vibration, and potentially future non-iPhone devices will support taptic feedback
            return (self.model.hasPrefix("iPhone")) || self.feedbackSupportLevel != .unsupported
        #endif
    }
    
    func vibrate()
    {
        guard self.isVibrationSupported else { return }
        
        switch self.feedbackSupportLevel
        {
        case .unsupported:
            AudioServicesStopSystemSound(kSystemSoundID_Vibrate)
            
            var vibrationLength = 30
            
            if self.modelGeneration.hasPrefix("iPhone6")
            {
                // iPhone 5S has a weaker vibration motor, so we vibrate for 10ms longer to compensate
                vibrationLength = 40;
            }
            
            // Must use NSArray/NSDictionary to prevent crash.
            let pattern: [Any] = [false, 0, true, vibrationLength]
            let dictionary: [String: Any] = ["VibePattern": pattern, "Intensity": 1]
            
            AudioServicesPlaySystemSoundWithVibration(kSystemSoundID_Vibrate, nil, dictionary as NSDictionary)
        
        case .basic, .feedbackGenerator: AudioServicesPlaySystemSound(1519) // "peek" vibration
        }
    }
}

private extension UIDevice
{
    var modelGeneration: String {
        var sysinfo = utsname()
        uname(&sysinfo)
        
        var modelGeneration: String!
        
        withUnsafePointer(to: &sysinfo.machine) { pointer in
            pointer.withMemoryRebound(to: UInt8.self, capacity: Int(Mirror(reflecting: pointer.pointee).children.count), { (pointer) in
                modelGeneration = String(cString: pointer)
            })
        }
        
        return modelGeneration
    }
}
